packages/cice5/package.py

```diff
@@ -0,0 +1,218 @@
+
+from spack.package import install, join_path, mkdirp
+
+# https://spack.readthedocs.io/en/latest/build_systems/makefilepackage.html
+class Cice5(MakefilePackage):
+    """The Los Alamos sea ice model (CICE) is the result of an effort to develop a computationally efficient sea ice component for a fully coupled atmosphere-land global climate model."""
+
+    homepage = "https://www.access-nri.org.au"
+    git = "https://github.com/ACCESS-NRI/cice5.git"
+
+    maintainers = ["harshula"]
+
+    version("spack-build", branch="spack-build")
+
+    patch(
+        "0001-Revert-Fix-Error-Period-required-in-format-specifier-E.patch",
+        when="%intel@2019.5.281"
+    )
+
+    # Depend on virtual package "mpi".
+    depends_on("mpi")
+    depends_on("oasis3-mct")
+    depends_on("datetime-fortran")
+    depends_on("netcdf-fortran@4.5.2:")
+    depends_on("netcdf-c@4.7.4:")
+    # TODO: For initial verification we are going to use static pio.
+    #       Eventually we plan to move to shared pio
+    # ~shared requires: https://github.com/spack/spack/pull/34837
+    depends_on("parallelio~pnetcdf~timing~shared")
+    depends_on("libaccessom2")
+
+    phases = ["edit", "build", "install"]
+
+    # The integr represents environment variable NTASK
+    __targets = {24: {}, 480: {}, 722: {}, 1682: {}}
+    __targets[24]["driver"] = "auscom"
+    __targets[24]["grid"] = "360x300"
+    __targets[24]["blocks"] = "24x1"
+
+    __targets[480]["driver"] = "auscom"
+    __targets[480]["grid"] = "1440x1080"
+    __targets[480]["blocks"] = "48x40"
+
+    # Comment from bld/config.nci.auscom.3600x2700:
+    # Recommendations:
+    #   use processor_shape = slenderX1 or slenderX2 in ice_in
+    #   one per processor with distribution_type='cartesian' or
+    #   squarish blocks with distribution_type='rake'
+    # If BLCKX (BLCKY) does not divide NXGLOB (NYGLOB) evenly, padding
+    # will be used on the right (top) of the grid.
+    __targets[722]["driver"] = "auscom"
+    __targets[722]["grid"] = "3600x2700"
+    __targets[722]["blocks"] = "90x90"
+
+    __targets[1682]["driver"] = "auscom"
+    __targets[1682]["grid"] = "3600x2700"
+    __targets[1682]["blocks"] = "200x180"
+
+    def edit(self, spec, prefix):
+
+        srcdir = self.stage.source_path
+        makeinc_path = join_path(srcdir, "bld", "Macros.spack")
+
+        config = {}
+
+        incs = (spec["oasis3-mct"].headers).cpp_flags + "/psmile.MPI1" + " "
+        for l in ["parallelio", "oasis3-mct", "libaccessom2", "netcdf-fortran"]:
+            incs += (spec[l].headers).cpp_flags + " "
+
+        # NOTE: The order of the libraries matter during the linking step!
+        # NOTE: datetime-fortran is a dependency of libaccessom2.
+        libs = "-L" + (spec["parallelio"].prefix) + "/lib -lpiof -lpioc "
+        for l in ["oasis3-mct", "libaccessom2", "netcdf-c", "netcdf-fortran", "datetime-fortran"]:
+            libs += (spec[l].libs).ld_flags + " "
```

Comment:
```suggestion
        libs = ["-L" + join_path((spec["parallelio"].prefix), "lib"), "-lpiof", "-lpioc"]
        libs +=  [(spec[l].libs).ld_flags for l in ["oasis3-mct", "libaccessom2", "netcdf-c", "netcdf-fortran", "datetime-fortran"]]
        libs = " ".join(libs)
```

Reply:
This avoids appending a space, avoids traversing the list twice and uses join():
```
        lstr = " ".join(
                    ["".join(["-L", join_path(spec["parallelio"].prefix, "lib")]),
                    "-lpiof",
                    "-lpioc"]
                   )
        ldeps = ["oasis3-mct", "libaccessom2", "netcdf-c", "netcdf-fortran", "datetime-fortran"]
        libs = " ".join([lstr] + [(spec[d].libs).ld_flags for d in ldeps])
```

Reply:
https://discuss.python.org/t/joining-strings-is-faster-than-adding-but-only-if-concatenating-3-strings/5293/2